packages/loopring_v3/contracts/converters/BaseConverter.sol

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+
+import "../core/iface/IExchangeV3.sol";
+import "../lib/Claimable.sol";
+import "../lib/Drainable.sol";
+import "../lib/ERC20.sol";
+import "../lib/AddressUtil.sol";
+import "../lib/ERC20SafeTransfer.sol";
+import "../lib/MathUint.sol";
+import "../lib/LPERC20.sol";
+
+
+/// @author Brecht Devos - <brecht@loopring.org>
+abstract contract BaseConverter is LPERC20, Claimable, Drainable
+{
+    using AddressUtil       for address;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+
+    event ConversionSuccess (uint amountIn, uint amountOut);
+    event ConversionFailed  (string reason);
+
+    IExchangeV3        public  immutable exchange;
+    IDepositContract   public  immutable depositContract;
+
+    bool               public  initialized;
+
+    address            public  tokenIn;
+    address            public  tokenOut;
+
+    bool               public  failed;
+
+    modifier onlyFromExchangeOwner()
+    {
+        require(msg.sender == exchange.owner(), "UNAUTHORIZED");
+        _;
+    }
+
+    constructor(
+        IExchangeV3 _exchange
+        )
+    {
+        exchange = _exchange;
+        depositContract = _exchange.getDepositContract();
+    }
+
+    function initialize(
+        string memory _name,
+        string memory _symbol,
+        uint8         _decimals,
```

Comment:
The reason I made it configurable is that the vault token is always minted 1:1 with the `tokenIn` token, so it would make sense if they have the same amount of decimals. But this of course has no impact on anything directly, except perhaps how we decide how the token is shown in the UI. If that always just looks at the contract it makes sense to set the "correct" value, otherwise a fixed value can indeed be used without any effect.